node/src/components/deploy_acceptor.rs

```diff
@@ -5,41 +5,126 @@ mod tests;
 
 use crate::{
     components::Component,
     effect::{
         announcements::DeployAcceptorAnnouncement,
         requests::{ContractRuntimeRequest, StorageRequest},
-        EffectBuilder, EffectExt, Effects,
+        EffectBuilder, EffectExt, Effects, Responder,
     },
     types::{
-        chainspec::DeployConfig, Chainspec, Deploy, DeployValidationFailure, NodeId, Timestamp,
+        chainspec::DeployConfig, Block, Chainspec, Deploy, DeployConfigurationFailure, NodeId,
+        Timestamp,
     },
     utils::Source,
     NodeRng,
 };
 
-use crate::effect::Responder;
 pub(crate) use config::Config;
-pub(crate) use event::Event;
+pub(crate) use event::{Event, EventMetadata};
+
+const ARG_TARGET: &str = "target";
+const ARG_SOURCE: &str = "source";
 
-#[derive(Debug, Error)]
+#[derive(Debug, Error, Serialize)]
+#[allow(clippy::enum_variant_names)]
 pub(crate) enum Error {
+    /// An invalid deploy was received from the client.
+    #[error("block chain has no blocks")]
+    InvalidBlockchain,
+
     /// An invalid deploy was received from the client.
     #[error("invalid deploy: {0}")]
-    InvalidDeploy(DeployValidationFailure),
-    /// An invalid account sent a deploy.
-    #[error("invalid account")]
-    InvalidAccount,
+    InvalidDeployConfiguration(DeployConfigurationFailure),
+
+    /// An invalid deploy was received from the client.
+    #[error("deploy parameter failure: {failure} at prestate_hash: {prestate_hash}")]
+    InvalidDeployParameters {
+        prestate_hash: Digest,
+        failure: DeployParameterFailure,
+    },
+}
+
+/// A representation of the way in which a deploy failed validation checks.
+#[derive(Clone, DataSize, Ord, PartialOrd, Eq, PartialEq, Hash, Debug, Error, Serialize)]
+pub enum DeployParameterFailure {
+    /// Account does not exist
+    #[error("Account does not exist")]
+    NonexistentAccount { account_hash: AccountHash },
+    /// Nonexistent contract at hash
+    #[error("Contract at {contract_hash} does not exist")]
+    NonexistentContractAtHash { contract_hash: ContractHash },
+    #[error("Contract named {name} does not exist in Account's NamedKeys")]
+    NonexistentContractAtName { name: String },
+    /// Nonexistent contract entrypoint
+    #[error("Contract does not have {entry_point}")]
+    NonexistentContractEntryPoint { entry_point: String },
+    /// Contract Package does not exist
+    #[error("Contract Package at {contract_package_hash} does not exist")]
+    NonexistentContractPackageAtHash {
+        contract_package_hash: ContractPackageHash,
+    },
+    #[error("ContractPackage named {name} does not exist in Account's NamedKeys")]
+    NonexistentContractPackageAtName { name: String },
+    /// Invalid contract at given version.
+    #[error("Invalid contract at version: {contract_version}")]
+    InvalidContractAtVersion { contract_version: ContractVersion },
+    /// Invalid associated keys
+    #[error("Account authorization invalid")]
+    InvalidAssociatedKeys,
```

Comment:
Why no account (hash) in this error variant?

Reply:
We report this error when the `approvals` on the `Deploy` aren't part of the account's associated keys. It would be more appropriate to have the `BTreeSet<AccountHash>`. However, I don't know if that is something we want to send over the wire given how large that set can be

Reply:
I meant an account hash for which that error was raised.